File: Human_written_code/Code_with_comments/Python/Q9.py
class Item:
    def __init__(self, value: int, weight: int):
        self.value = value   # Value of the item
        self.weight = weight # Weight of the item

class Solution:
    def fractionalKnapsack(self, W: int, arr: List[Item], n: int) -> float:
        # Sort items based on their value-to-weight ratio in descending order
        arr.sort(key=lambda x: x.value / x.weight, reverse=True)
        
        curWeight = 0        # Current weight in the knapsack
        finalValue = 0.0     # Final value of the knapsack
        
        # Iterate through the sorted items
        for i in range(n):
            # If adding the full item does not exceed the capacity
            if curWeight + arr[i].weight <= W:
                curWeight += arr[i].weight  # Add item's weight
                finalValue += arr[i].value  # Add item's value
            else:
                # If adding the full item exceeds the capacity, add fraction of it
                remain = W - curWeight  # Calculate remaining capacity
                finalValue += arr[i].value / arr[i].weight * remain  # Add fractional value
                break  # Knapsack is now full
        
        return finalValue  # Return the maximum value that can be obtained

if __name__ == '__main__':
    n = 3  # Number of items
    W = 50 # Capacity of the knapsack
    # List of items with their value and weight
    arr = [Item(60, 10), Item(100, 20), Item(120, 30)]
    
    obj = Solution()
    ans = obj.fractionalKnapsack(W, arr, n)
    
    print("The maximum value is", ans)  # Print the maximum value

File: Human_written_code/Code_with_comments/CPP/Q9.cpp
#include <bits/stdc++.h>

using namespace std;

struct Item {
   int value;
   int weight;
};

class Solution {
   public:
      // Comparator to sort items by value-to-weight ratio
      bool static comp(Item a, Item b) {
         double r1 = (double) a.value / (double) a.weight;
         double r2 = (double) b.value / (double) b.weight;
         return r1 > r2;
      }

   // Function to return the maximum value in the knapsack
   double fractionalKnapsack(int W, Item arr[], int n) {
      sort(arr, arr + n, comp);  // Sort items by value-to-weight ratio

      int curWeight = 0;
      double finalvalue = 0.0;

      // Loop through sorted items and add full or fractional parts to the knapsack
      for (int i = 0; i < n; i++) {
         if (curWeight + arr[i].weight <= W) {
            curWeight += arr[i].weight;
            finalvalue += arr[i].value;
         } else {
            int remain = W - curWeight;
            finalvalue += (arr[i].value / (double) arr[i].weight) * (double) remain;
            break; // Stop when the knapsack is full
         }
      }

      return finalvalue;
   }
};

int main() {
   int n = 3, weight = 50;
   Item arr[n] = { {100, 20}, {60, 10}, {120, 30} };
   Solution obj;
   double ans = obj.fractionalKnapsack(weight, arr, n);
   cout << "The maximum value is " << setprecision(2) << fixed << ans;
   return 0;
}
